### PR TITLE
kubelet: fix PID based eviction

### LINKED ISSUE #107063: Tests on sig-node-containerd#node-kubelet-containerd-eviction are failing
### Which jobs are failing?

containerd eviction tests

### Which tests are failing?

```
E2eNode Suite.[sig-node] LocalStorageEviction [Slow] [Serial] [Disruptive][NodeFeature:Eviction] when we run containers that should cause DiskPressure should eventually evict all of the correct podsE2eNode Suite.[sig-node] LocalStorageSoftEviction [Slow] [Serial] [Disruptive][NodeFeature:Eviction] when we run containers that should cause DiskPressure should eventually evict all of the correct pods
```

### Since when has it been failing?

12/10, seems like since ported from dockershim 

### Testgrid link

https://testgrid.k8s.io/sig-node-containerd#node-kubelet-containerd-eviction

### Reason for failure (if possible)

_No response_

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig node
/priority important-soon

### LINKED ISSUE #115215: kubelet: Process Stats are always 0 as returned by kubelet summary API
### What happened?

[Process stats](https://pkg.go.dev/k8s.io/kubelet/pkg/apis/stats/v1alpha1#ProcessStats) are not currently returned in kubelet summary API.

### What did you expect to happen?

[Process stats](https://pkg.go.dev/k8s.io/kubelet/pkg/apis/stats/v1alpha1#ProcessStats) are returned correctly in kubelet summary API.

### How can we reproduce it (as minimally and precisely as possible)?

1. Start a cluster (using kind cluster from master)

```
$ kubectl get nodes
NAME                 STATUS   ROLES           AGE     VERSION
kind-control-plane   Ready    control-plane   7m49s   v1.27.0-alpha.0.1120+56a14025b06734
kind-worker          Ready    <none>          7m27s   v1.27.0-alpha.0.1120+56a14025b06734
```

3. Launch a pod and consume a few pids, e.g. 10 pids
4. Fetch summary api and examine the process stats

```
$ kubectl run -i --tty --rm debug --image=k8s.gcr.io/busybox:latest --restart=Never -- /bin/sh
/ # for i in $(seq 1 10); do sleep 99999999 & done
/ # ps aux
PID   USER     COMMAND
    1 root     /bin/sh
   16 root     sleep 99999999
   17 root     sleep 99999999
   18 root     sleep 99999999
   19 root     sleep 99999999
   20 root     sleep 99999999
   21 root     sleep 99999999
   22 root     sleep 99999999
   23 root     sleep 99999999
   24 root     sleep 99999999
   25 root     sleep 99999999
   28 root     ps aux
/ #


$ kubectl get --raw "/api/v1/nodes/kind-worker/proxy/stats/summary"


  {
   "podRef": {
    "name": "debug",
    "namespace": "default",
    "uid": "4b58edc1-917a-4273-8e30-113e65119ba6"
   },
   "startTime": "2023-01-19T22:03:01Z",
   "containers": [
    {
     "name": "debug",
     "startTime": "2023-01-19T22:03:02Z",
     "cpu": {
      "time": "2023-01-19T22:06:49Z",
      "usageNanoCores": 0,
      "usageCoreNanoSeconds": 108558000
     },
     "memory": {
      "time": "2023-01-19T22:06:49Z",
      "usageBytes": 1925120,
      "workingSetBytes": 1912832,
      "rssBytes": 716800,
      "pageFaults": 2909,
      "majorPageFaults": 0
     },
     "rootfs": {
      "time": "2023-01-19T22:07:03Z",
      "availableBytes": 47482068992,
      "capacityBytes": 431596691456,
      "usedBytes": 36864,
      "inodesFree": 21663417,
      "inodes": 26836992,
      "inodesUsed": 10
     },
     "logs": {
      "time": "2023-01-19T22:07:08Z",
      "availableBytes": 47482068992,
      "capacityBytes": 431596691456,
      "usedBytes": 8192,
      "inodesFree": 21663417,
      "inodes": 26836992,
      "inodesUsed": 1
     }
    }
   ],
   "cpu": {
    "time": "2023-01-19T22:06:59Z",
    "usageNanoCores": 0,
    "usageCoreNanoSeconds": 149739000
   },
   "memory": {
    "time": "2023-01-19T22:06:59Z",
    "usageBytes": 2244608,
    "workingSetBytes": 2232320,
    "rssBytes": 757760,
    "pageFaults": 3991,
    "majorPageFaults": 0
   },
   "network": {
    "time": "2023-01-19T22:07:04Z",
    "name": "eth0",
    "rxBytes": 0,
    "rxErrors": 0,
    "txBytes": 866,
    "txErrors": 0,
    "interfaces": [
     {
      "name": "eth0",
      "rxBytes": 0,
      "rxErrors": 0,
      "txBytes": 866,
      "txErrors": 0
     }
    ]
   },
   "volume": [
    {
     "time": "2023-01-19T22:05:45Z",
     "availableBytes": 67012878336,
     "capacityBytes": 67012890624,
     "usedBytes": 12288,
     "inodesFree": 8180275,
     "inodes": 8180284,
     "inodesUsed": 9,
     "name": "kube-api-access-pqck4"
    }
   ],
   "ephemeral-storage": {
    "time": "2023-01-19T22:07:08Z",
    "availableBytes": 47482068992,
    "capacityBytes": 431596691456,
    "usedBytes": 49152,
    "inodesFree": 21663417,
    "inodes": 26836992,
    "inodesUsed": 12
   },
   "process_stats": {
    "process_count": 0 <----------- This is reported as 0 incorrectly !!!
   }
  }

```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind bug
#### What this PR does / why we need it:
carry https://github.com/kubernetes/kubernetes/pull/123369 plus an additional patch to fix cadvisor stats provider
#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/115215 
Fixes https://github.com/kubernetes/kubernetes/issues/107063

#### Does this PR introduce a user-facing change?

```release-note
Fix a bug where the Kubelet didn't calculate the process usage of pods correctly, leading to pods never getting evicted for PID use.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by haircommander** [comment]
untested so far :yolo:

/test pull-crio-cgroupv2-node-e2e-eviction
/test pull-crio-cgroupv1-node-e2e-eviction

**Comment by kannon92** [comment]
PR is looking good!

**Comment by haircommander** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124101/pull-crio-cgroupv1-node-e2e-eviction/1773400529385295872 :green_circle: :tada:

**Comment by kannon92** [comment]
@haircommander can you add an useful release note?

**Comment by haircommander** [comment]
> @haircommander can you add an useful release note?

done!

File: pkg/kubelet/stats/cadvisor_stats_provider.go [hunk 1]
```diff
@@ -138,6 +138,7 @@ func (p *cadvisorStatsProvider) ListPodStats(_ context.Context) ([]statsapi.PodS
 				containerStat.Logs = logs
 			}
 			podStats.Containers = append(podStats.Containers, *containerStat)
+			podStats.ProcessStats = mergeProcessStats(podStats.ProcessStats, cadvisorInfoToProcessStats(&cinfo))
```
Comment (kannon92) [inline pkg/kubelet/stats/cadvisor_stats_provider.go:141]: Maybe you could leave a comment here to mention why we are computing processes at the container level.

**Comment by kannon92** [comment]
one small nit. https://github.com/kubernetes/kubernetes/pull/124101/files#r1544717415

otherwise,
/lgtm

**Comment by kannon92** [comment]
cc @pacoxu @mrunalp @SergeyKanzhelev @dchen1107

**Comment by kannon92** [comment]
/triage accepted
/priority important-soon

We had this bug for long enough so I hope we can get in as soon as main opens up.

**Comment by kannon92** [comment]
@SergeyKanzhelev do you think this would be worth a cherry pick?

File: pkg/kubelet/stats/cadvisor_stats_provider.go [hunk 2]
```diff
@@ -154,7 +155,7 @@ func (p *cadvisorStatsProvider) ListPodStats(_ context.Context) ([]statsapi.PodS
 			podStats.CPU = cpu
 			podStats.Memory = memory
 			podStats.Swap = cadvisorInfoToSwapStats(podInfo)
-			podStats.ProcessStats = cadvisorInfoToProcessStats(podInfo)
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/stats/cadvisor_stats_provider.go:157]: will the cAdvisor still collect it? Should we make it not to collect it?  Would it make sense to merge ProcessStats with the pause process stats?

**Comment by kannon92** [comment]
@haircommander do you think it would be possible to get this in?

File: pkg/kubelet/stats/helper.go [hunk 3]
```diff
@@ -168,6 +168,31 @@ func cadvisorInfoToProcessStats(info *cadvisorapiv2.ContainerInfo) *statsapi.Pro
...
+
+	if first == nil {
+		return second
+	}
+	if second == nil {
+		return first
+	}
+
+	firstProcessCount := uint64(0)
+	if first != nil && first.ProcessCount != nil {
```
Comment (HirazawaUi) [inline pkg/kubelet/stats/helper.go:184]: here is no need to check if `first` is nil.

Comment (haircommander) [reply pkg/kubelet/stats/helper.go]: great point! fixed with second as well

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

this looks good to me

**Comment by kannon92** [comment]
@haircommander can we see about getting this in? We got approval from @SergeyKanzhelev but I guess we missed a few approvals.

**Comment by haircommander** [comment]
yeah @SergeyKanzhelev @mrunalp do you think this could be added to the milestone and approved?

**Comment by kannon92** [comment]
@dims @dchen1107 can you help here? This PR should fix some eviction test failures.

**Review by mrunalp — APPROVED** [review]

**Comment by neoaggelos** [comment]
Hi folks! We are around 2 days past code freeze, would be so kind as to file a code freeze exception request for this PR?

cc @katcosgrove @gracenng

**Comment by kannon92** [comment]
I created an exception for this. Its not a KEP but fixing existing tests.

**Pull request merged**